File: 2024/JS/Day5/solutionB.js
const fs = require("fs");

function sortPages(pages, rules) {
  const graph = new Map();
  const inDegree = new Map();

  for (const page of pages) {
    graph.set(page, []);
    inDegree.set(page, 0);
  }

  for (const [x, y] of rules) {
    if (graph.has(x) && graph.has(y)) {
      graph.get(x).push(y);
      inDegree.set(y, inDegree.get(y) + 1);
    }
  }

  const queue = [];
  for (const [node, degree] of inDegree) {
    if (degree === 0) queue.push(node);
  }

  const sorted = [];
  while (queue.length > 0) {
    const current = queue.shift();
    sorted.push(current);

    for (const neighbor of graph.get(current)) {
      inDegree.set(neighbor, inDegree.get(neighbor) - 1);
      if (inDegree.get(neighbor) === 0) queue.push(neighbor);
    }
  }

  return sorted;
}

function isUpdateValid(update, rules) {
  const pageIndex = new Map();
  update.forEach((page, idx) => pageIndex.set(page, idx));

  for (const [x, y] of rules) {
    if (
      pageIndex.has(x) &&
      pageIndex.has(y) &&
      pageIndex.get(x) > pageIndex.get(y)
    ) {
      return false;
    }
  }
  return true;
}

function findMiddlePage(update) {
  const middleIndex = Math.floor(update.length / 2);
  return update[middleIndex];
}

function solveChallenge(input) {
  const [rulesSection, updatesSection] = input.trim().split("\n\n");
  const rules = rulesSection
    .split("\n")
    .map((rule) => rule.split("|").map(Number));
  const updates = updatesSection
    .split("\n")
    .map((update) => update.split(",").map(Number));

  let total = 0;

  for (const update of updates) {
    if (!isUpdateValid(update, rules)) {
      // Reorder the incorrect update
      const sortedUpdate = sortPages(update, rules);
      total += findMiddlePage(sortedUpdate);
    }
  }

  return total;
}

fs.readFile("../../data/day5.txt", (err, data) => {
  let input = data.toString();
  console.log(solveChallenge(input));
});

File: 2024/Python/Day5/solutionB.py
from collections import defaultdict, deque

def parse_input(input_data):
    rules_section, updates_section = input_data.strip().split("\n\n")
    rules = [tuple(map(int, line.split("|"))) for line in rules_section.splitlines()]
    updates = [list(map(int, line.split(","))) for line in updates_section.splitlines()]
    return rules, updates

def is_update_valid(update, rules):
    page_indices = {page: idx for idx, page in enumerate(update)}
    for x, y in rules:
        if x in page_indices and y in page_indices:
            if page_indices[x] > page_indices[y]:
                return False
    return True

def sort_pages(update, rules):
    graph = defaultdict(list)
    in_degree = defaultdict(int)

    for page in update:
        graph[page] = []
        in_degree[page] = 0

    # Create graph and calculate in-degrees
    for x, y in rules:
        if x in graph and y in graph:
            graph[x].append(y)
            in_degree[y] = in_degree.get(y, 0) + 1

    # Restrict graph to only the pages in this update
    valid_nodes = set(update)
    graph = {k: [v for v in graph[k] if v in valid_nodes] for k in valid_nodes}
    in_degree = {k: in_degree[k] for k in valid_nodes}

    # Queue of nodes with zero in-degree
    queue = deque([node for node in valid_nodes if in_degree[node] == 0])
    sorted_list = []

    while queue:
        node = queue.popleft()
        sorted_list.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)
    return sorted_list

def reorder_update(update, rules):
    if not update:  # Handle empty updates explicitly
        return []
    return sort_pages(update, rules)

def find_middle_page(update):
    return update[len(update) // 2]

def solve_challenge(input_data):
    rules, updates = parse_input(input_data)
    total = 0

    for update in updates:
        if not is_update_valid(update, rules):  # If invalid, reorder the update
            reordered_update = reorder_update(update, rules)
            if reordered_update:  # Ensure the reordered update is non-empty
                total += find_middle_page(reordered_update)

    return total

text_input = ""
with open("../../data/day5.txt") as file:
    for line in file:
        text_input += line

result = solve_challenge(text_input)
print(result)
